### PR TITLE
plat-ls: updated conf.mk to set CFG_USER_TA_TARGETS ?= ta_arm64

### PR DESCRIPTION
**Pull request by pangupta**
Signed-off-by: Pankaj Gupta <pankaj.gupta@nxp.com>

### CODE REVIEW

**Comment by jforissier** [comment]
Hi @pangupta,

About the Shippable CI error:
The current line (`_make PLATFORM=ls-ls1021atwr`) performs a 32-bit build, which is the default.
If your goal is to create a full 64-bit build by default (TEE core and TA being 64-bits), you need to set `CFG_ARM64_core ?= y`, too. Indeed, 64-bit TAs are supported only when the TEE core is 64-bits.

**Comment by pangupta** [comment]
> Hi @pangupta,
> 
> About the Shippable CI error:
> The current line (`_make PLATFORM=ls-ls1021atwr`) performs a 32-bit build, which is the default.
> If your goal is to create a full 64-bit build by default (TEE core and TA being 64-bits), you need to set `CFG_ARM64_core ?= y`, too. Indeed, 64-bit TAs are supported only when the TEE core is 64-bits.

Setting CFG_ARM64_core = y only will not help. From R3.4.0 onwards, CFG_USER_TA_TARGETS cannot be left blank. 
If it is not set, default values of this flag is "ta_arm64 for ta_arm32".
This means, for a particular 64bit platform, if OPTEE-OS is complied, then it will try to compile 32-bit user TAs as well. 
In that case, if CROSS_COMPILE32 is not set, the compilation halted with errors.

To fix this, I have raised this pull request.

File: core/arch/arm/plat-ls/conf.mk [hunk 1]
```diff
@@ -103,12 +103,14 @@ CFG_TZDRAM_SIZE ?= ( CFG_TEE_OS_DRAM0_SIZE - CFG_SHMEM_SIZE)
...
 CFG_SHMEM_START ?= ((CFG_DRAM0_BASE + CFG_DRAM0_SIZE) - CFG_SHMEM_SIZE)
 $(call force,CFG_ARM64_core,y)
+CFG_USER_TA_TARGETS ?= ta_arm64
 else
 #In ARMv7 platform CFG_SHMEM_SIZE is different to that of ARMv8 platforms.
 CFG_TZDRAM_START ?= ((CFG_DRAM0_BASE + CFG_DRAM0_SIZE) - CFG_TEE_OS_DRAM0_SIZE)
 CFG_TZDRAM_SIZE ?= ( CFG_TEE_OS_DRAM0_SIZE - (2*CFG_SHMEM_SIZE))
 #CFG_SHMEM_START (Non-Secure shared memory) needs to be 2MB aligned boundary for TZASC 380 configuration.
 CFG_SHMEM_START ?= ((CFG_DRAM0_BASE + CFG_DRAM0_SIZE) - (2*CFG_SHMEM_SIZE))
+CFG_USER_TA_TARGETS ?= ta_arm32
```
Comment (jforissier) [inline core/arch/arm/plat-ls/conf.mk]: Is this needed? This part is for 32-bit TEE core, `ta_arm32` should be the default for `CFG_USER_TA_TARGETS`.

**Comment by jforissier** [comment]
I don't understand what you are trying to fix here. `CROSS_COMPILE32` defaults to `$(CROSS_COMPILE)` which defaults to `arm-linux-gnueabihf-`. So it cannot be "not set". How are you calling `make`?

The current configuration can build a 64-bit core + dual-arch TA configuration without problem:
```
$ make -j10 -s PLATFORM=ls-ls1012ardb
$ file out/arm-plat-ls/core/tee.elf
out/arm-plat-ls/core/tee.elf: ELF 64-bit LSB executable, ARM aarch64, version 1 (SYSV), statically linked, not stripped
$ ls -1 out/arm-plat-ls/ta*-lib/libutee/libutee.a
out/arm-plat-ls/ta_arm32-lib/libutee/libutee.a
out/arm-plat-ls/ta_arm64-lib/libutee/libutee.a
```

Now, it's your call to choose not to build 32-bit TA libs by default, but it's not a fix?

**Comment by pangupta** [comment]
Let me share you the issue I am facing:

If you run these two commands back to back:
$> export CROSS_COMPILE=aarch64-linux-gnu-
$> make PLATFORM=ls-ls1046ardb ARCH=arm

Following error comes which were not coming earlier:
  CC      out/arm-plat-ls/ta_arm32-lib/libmbedtls/mbedtls/library/aes.o
aarch64-linux-gnu-gcc: error: unrecognized command line option -mthumb
aarch64-linux-gnu-gcc: error: unrecognized command line option -mthumb-interwork
aarch64-linux-gnu-gcc: error: unrecognized command line option -mno-unaligned-access; did you mean -Wno-aligned-new?
aarch64-linux-gnu-gcc: error: unrecognized command line option -mfloat-abi=hard
mk/compile.mk:147: recipe for target 'out/arm-plat-ls/ta_arm32-lib/libmbedtls/mbedtls/library/aes.o' failed
make: *** [out/arm-plat-ls/ta_arm32-lib/libmbedtls/mbedtls/library/aes.o] Error 1


This is the case where user either do not set or sets CFG_USER_TA_TARGETS="ta_arm32 ta_arm64", in order to compile TAs for bot the modes, then compilation will fail.

This error is coming because, on setting CROSS_COMPILE=<64-bit toolchain> leads to CROSS_COMPILE32 points to 64 bit tool chain as well.

I am trying to fix this by not compiling TAs (by default) for both modes.

Please suggest your view.

**Comment by jforissier** [comment]
`CROSS_COMPILE` is reserved for the 32-bit cross compiler. Just remove `export CROSS_COMPILE=aarch64-linux-gnu-` and it should build properly. And btw ARCH=arm is useless.

File: core/arch/arm/plat-ls/conf.mk (hunk 1 above)
Comment (pangupta) [reply core/arch/arm/plat-ls/conf.mk]: > Is this needed? This part is for 32-bit TEE core, `ta_arm32` should be the default for `CFG_USER_TA_TARGETS`.

I have removed the line from else part.

> Now, it's your call to choose not to build 32-bit TA libs by default, but it's not a fix?

For NXP build environment, CROSS_COMPILE is set to 32-bit or 64-bit toolchain depending on the platform.

To accommodate the  release upgrade of the OPTEE-OS , in the existing NXP build frame-work, this change is important.

Thanks for your review and replies.
Raising the new pull request with above changes.

**Comment by jforissier** [comment]
Thanks for making the commit description consistent with the change.
`Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

**Comment by pangupta** [comment]
Please plan to merge this.
Thanks.

**Comment by jforissier** [comment]
Please apply my A-b tag, then I can merge this. Thanks.

**Pull request merged**